Sources/SwiftSyntax/SyntaxVisitor.swift.gyb

```diff
@@ -141,3 +141,75 @@ open class SyntaxVisitor {
     }
   }
 }
+
+public protocol SyntaxTransformVisitor {
+  associatedtype ResultType = Void
+  
+  func visit(_ token: TokenSyntax) -> [ResultType]
+  func visit(_ node: UnknownSyntax) -> [ResultType]
+  
+% for node in SYNTAX_NODES:
+%   if is_visitable(node):
+  /// Visiting `${node.name}` specifically.
+  ///   - Parameter node: the node we are visiting.
+  ///   - Returns: the sum of whatever the child visitors return.
+  func visit(_ node: ${node.name}) -> [ResultType]
+%   end
+% end
+}
+
+extension SyntaxTransformVisitor {
+  public func visit(_ token: TokenSyntax) -> [ResultType] { [] }
+  public func visit(_ node: UnknownSyntax) -> [ResultType] { [] }
+
+% for node in SYNTAX_NODES:
+%   if is_visitable(node):
+  /// Visiting `${node.name}` specifically.
+  ///   - Parameter node: the node we are visiting.
+  ///   - Returns: nil by default.
+  public func visit(_ node: ${node.name}) -> [ResultType] {
+%   if node.is_base():
+      // Avoid calling into visitChildren if possible.
+      if !node.raw.layoutView!.children.isEmpty {
+        return visitChildren(node)
+      }
+      return []
+%   else:
+      // Avoid calling into visitChildren if possible.
+      if !node.raw.layoutView!.children.isEmpty {
+        return visitChildren(node)
+      }
+      return []
+%   end
+  }
+%   end
+% end
+
+  public func visit(_ data: SyntaxData) -> [ResultType] {
```

Comment:
We don't want clients visiting SyntaxData directly.

Reply:
I agree, `SyntaxTransformVisitor` should be a `class` and this should be `private` like we do in all the other initializers.

Reply:
Why do we want this to be a class? I think this should just take a `Syntax`.